File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    // time - O(n * 2^n), space - O(2^n)
    combinationSum2(candidates, target) {
        candidates.sort((a,b) => a-b);
        const combinations = [];
        const recursion = (index, remaining, nums) => {
            if (remaining === 0) {
                combinations.push([...nums])
                return
            }
            if (index === candidates.length || remaining < 0) return
            const num = candidates[index]
            nums.push(num)
            recursion(index+1, remaining-num, nums)
            nums.pop()
            while (index < candidates.length && candidates[index] === candidates[index+1]) {
                index++
            }
            recursion(index+1, remaining, nums)
        }
        recursion(0, target, [])
        return combinations
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {
    // time - O(n^(t/m)), space - O(t/m) - t is target and m is min value in candidates, n is length of candidates
    private List<List<Integer>> output = new ArrayList<>();
    private List<Integer> combination = new ArrayList<>();

    private void backtrack(int index, int[] candidates, int remaining) {
        if (remaining == 0) {
            output.add(new ArrayList<>(combination));
            return;
        }
        for (int i = index; i < candidates.length; i++) {
            int diff = remaining - candidates[i];
            if (diff < 0) {
                break;
            }
            combination.add(candidates[i]);
            backtrack(i, candidates, diff);
            combination.remove(combination.size() - 1);
        }

    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtrack(0, candidates, target);
        return output;
    }
}